File: A1/producer-consumer/producer-consumer.go
package main
/* producer-consumer problem in Go
*  from http://www.golangpatterns.info/concurrency/producer-consumer
*/

import (
    "fmt"
    "time"
    "math/rand"
    "github.com/shirou/gopsutil/cpu"
    "github.com/shirou/gopsutil/mem"
    "os"
)

const MAX_NUM = 10

var done = make(chan bool)
var msgs = make(chan int, MAX_NUM)
var nums = 5
var psum = 0
var csum = 0

func produce () {
    for {
      num := rand.Intn(nums)
      msgs <- num
      psum += 1
      //fmt.Printf("Produced token %v : it's a living!\n", num)
    }
}

func consume () {
    for {
      <-msgs
      csum += 1
      //msg := <-msgs
      //fmt.Printf("Consumed token %v : yum yum yum!\n", msg)
   }
}

func monitor () {
    iterations := 10
    CPUusage := 0.0
    VMusage := 0.0
    for i := 0; i < iterations; i++ {
      c, err := cpu.Percent(time.Duration(200) * time.Millisecond, false)
      m, err := mem.VirtualMemory()
      if err == nil {
        cpu := c[0]
        vm := m.UsedPercent
        CPUusage += cpu
        VMusage += vm
        fmt.Printf("CPU usage: %f\n", cpu)
        fmt.Printf("%v\n", m)
      }
      time.Sleep(time.Second)
    }
    CPUavg := CPUusage / float64(iterations)
    fmt.Printf("Average CPU usage: %f \n", CPUavg)
    VMavg := VMusage / float64(iterations)
    fmt.Printf("Average VM usage: %f \n", VMavg)
    fmt.Printf("Total productions: %d\tTotal consumptions: %d\n", psum, csum)
    os.Exit(0)
}

func main () {

    go monitor()
    time.Sleep(time.Second)

    go produce()
    go consume()
    <- done
}

File: A1/producer-consumer/producer-consumer.py
"""
producer-consumer problem in Python
from https://www.agiliq.com/blog/2013/10/producer-consumer-problem-in-python/
"""

from threading import Thread, Condition
import thread
import time
import random
import psutil

queue = []
MAX_NUM = 10
condition = Condition()
psum = 0
csum = 0

class ProducerThread(Thread):
    def run(self):
        nums = range(5)
        global queue
        global psum
        while True:
            condition.acquire()
            if len(queue) == MAX_NUM:
                #print "Queue full, producer is waiting"
                condition.wait()
                #print "Space in queue, Consumer notified the producer"
            num = random.choice(nums)
            queue.append(num)
            #print "Produced", num
            psum += 1
            condition.notify()
            condition.release()

class ConsumerThread(Thread):
    def run(self):
        global queue
        global csum
        while True:
            condition.acquire()
            if not queue:
                #print "Nothing in queue, consumer is waiting"
                condition.wait()
                #print "Producer added something to queue and notified the consumer"
            num = queue.pop(0)
            #print "Consumed", num
            csum += 1
            condition.notify()
            condition.release()

class Monitor(Thread):
    def run(self):
        iterations = 10
        CPUusage = []
        VMusage = []
        for i in range(iterations):
            CPU = psutil.cpu_percent(0.2, False)
            print "CPU usage:", CPU
            CPUusage.append(CPU)
            VM = psutil.virtual_memory()
            print VM
            VMusage.append(VM[2])
            time.sleep(1)
        CPUavg = sum(CPUusage)/float(len(CPUusage))
        print "Average CPU usage:", CPUavg
        VMavg = sum(VMusage)/float(len(VMusage))
        print "Average VM usage:", VMavg
        print "Total productions:", psum, "Total consumptions:", csum
        thread.interrupt_main()

Monitor().start()
time.sleep(1)

ProducerThread().start()
ConsumerThread().start()
